File: LeetCode-Python/ValidParentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        closing = {')': '(', ']': '[', '}': '{'}

        for char in s:
            if char in closing:
                if not stack or stack.pop() != closing[char]:
                    return False
            else:
                stack.append(char)

        return not stack

File: LeetCode-Cpp/ValidParentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        stack<char> st;
        unordered_map<char, char> closing = {{')', '('}, {']', '['}, {'}', '{'}};

        for (char c : s) {
            if (closing.find(c) != closing.end()) {
                if (st.empty() || st.top() != closing[c]) {
                    return false;
                }
                st.pop();
            } else {
                st.push(c);
            }
        }

        return st.empty();
    }
};
